### PR TITLE
Refine interface for Test...Utxo.genTx hard errors

### LINKED ISSUE #2557: RealTPraos ThreadNet nightly tests failing with genTx: insufficient spending balance
First appeared in 27 August 2020's https://buildkite.com/input-output-hk/ouroboros-network-nightly/builds/212#b8115662-22f5-4d36-a544-a4a1d411e742

Timing and locality suggest that this is indirectly due to #2550. #2554 may somehow related.

Edit: still present as of https://buildkite.com/input-output-hk/ouroboros-network-nightly/builds/217

### PR DESCRIPTION
**Pull request by nfrisby**
I include how close the balance managed to come to the fees in the exception because that might help indicate what went wrong the next time someone is debugging in this area.

I plan to use this to resolve input-output-hk/ouroboros-network#2557.

### CODE REVIEW

**Comment by nc6** [comment]
Just rebased this on current master, since I merged a PR that changed a lot!

File: shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs [hunk 1]
```diff
@@ -120,18 +124,49 @@ genTx ::
...
+  deriving (Show)
+
+-- | 'Left' means the generator could not afford the fees for any of the
+-- 'genTxRetries'-many transactions it attempted.
+tryGenTx ::
+  (HasCallStack, Era era, Mock (Crypto era)) =>
+  GenEnv era ->
+  LedgerEnv ->
+  (UTxOState era, DPState era) ->
+  Gen (Either InsufficientSpendingBalanceInfo (Tx era))
```
Comment (nc6) [inline shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs]: I feel if we have the `GenTxException` type then we should throw that. Seems weird to wrap it at the level of `genTx` whereas `tryGenTx` returns only one specific type.

**Review by nc6 — APPROVED** [review]
The meat of the change looks good, though I find the ergonomics a little unsatisfying - weird mix of pretending we have a generalised exception but most of the functions, though general, looking like they're specific to one exception type.

File: shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs (hunk 1 above)
Comment (nfrisby) [reply shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs]: Done. Thanks!

File: shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs [hunk 2]
```diff
@@ -120,18 +124,47 @@ genTx ::
...
   (UTxOState era, DPState era) ->
   Gen (Tx era)
-genTx ge@(GenEnv _ (Constants {genTxRetries})) =
-  genTxRetry genTxRetries ge
+genTx ge le st =
+  tryGenTx ge le st >>= \case
+    Left exn -> Exn.throw exn
+    Right tx -> pure tx
+
+data GenTxException
```
Comment (mrBliss) [inline shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs:132]: It's called "-Exception", but it isn't an instance of the `Exception` class and thus never actually "thrown". I typically use `-Error` for that.

Comment (nc6) [reply shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs:132]: "Alternative exit status"

File: shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs [hunk 3]
```diff
@@ -200,11 +233,18 @@ genTxRetry
...
-    where
-      retryOrFail 0 = error "genTx: insufficient spending balance"
-      retryOrFail n' = genTxRetry (n' - 1) ge env st
+          pure $ Right $ Tx txBody (mkTxWits' txBody) metadata
+        else do
+          let minDiffSoFar' = min minDiffSoFar (fees - spendingBalance)
+          if n < 1
+            then
+              pure $
+                Left
```
Comment (uroboros) [inline shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs]:
```suggestion
                Left $
```

**Review by uroboros — APPROVED** [review]
@TimSheard @redxaxder keep this PR in mind for your pending `genTx` refactor

File: shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs (hunk 2 above)
Comment (nfrisby) [reply shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs:132]: It is an instance (see just below this thread in the diff) and it is thrown (see just above this thread). Or am I confused about your observation?

**Pull request merged**